File: test-suite/handwritten-src/ts/DataTest.ts
import {TestCase, allTests, assertArrayEq} from "./testutils"
import * as test from "../../generated-src/ts/test";
import {DjinniModule} from "@djinni_support/DjinniModule"

class DataTest extends TestCase {
    m: test.Test_statics & DjinniModule;
    test: test.DataRefTest;

    constructor(module: test.Test_statics & DjinniModule) {
        super(module);
        this.m = module;
        this.test = this.m.testsuite.DataRefTest.create()!;
    }

    testSend() {
        var input = [0, 1, 2, 3];
        var buf = this.m.allocateWasmBuffer(4);
        buf.set(input);
        this.test.sendData(buf);
        var output = this.test.retriveAsBin();
        assertArrayEq(input, output);
    }

    testSendMutable() {
        var input = [0, 1, 2, 3];
        var buf = this.m.allocateWasmBuffer(4);
        buf.set(input);
        this.test.sendMutableData(buf);
        assertArrayEq([3, 2, 1, 0], buf);
    }

    testGenerateData() {
        var buf = this.test.generateData();
        assertArrayEq([0, 1, 2, 3], buf);
    };

    testDataFromVec() {
        var buf = this.test.dataFromVec();
        assertArrayEq([0, 1, 2, 3], buf);
    }

    testDataFromStr() {
        var buf = this.test.dataFromStr();
        // codepoints of a, b, c, d
        assertArrayEq([97, 98, 99, 100], buf);
    }

    testSendDataView() {
        var input = [0, 1, 2, 3];
        var buf = this.m.allocateWasmBuffer(4);
        buf.set(input);
        var output = this.test.sendDataView(buf);
        assertArrayEq(input, output);
    }
    
    testRecvDataView() {
        var output = this.test.recvDataView();
        assertArrayEq([0, 1, 2, 3], output);
    }
}

allTests.push(DataTest);

File: test-suite/handwritten-src/swift/DataTest.swift
import XCTest
import Foundation
@testable import TestSuite

final class DataTest: XCTestCase {

    let test = (try? DataRefTest_statics.create())!

    func testSend() throws {
        let input: [UInt8] = [0, 1, 2, 3]
        let buf = Data(input)
        try test.sendData(buf as NSData)
        let output = try test.retriveAsBin()
        XCTAssertEqual(Data(input), output)
    }

    func testSendMutable() throws {
        let input: [UInt8] = [0, 1, 2, 3]
        let buf = NSMutableData()
        buf.append(Data(input))
        try test.sendMutableData(buf)
        XCTAssertEqual(Data([3,2,1,0]), buf as Data)
    }

    func testGenerateData() throws {
        let buf = try test.generateData()
        let output = buf as Data
        XCTAssertEqual(Data([0, 1, 2, 3]), output)
    }

    func testDataFromVec() throws {
        let buf = try test.dataFromVec()
        let output = buf as Data
        XCTAssertEqual(Data([0, 1, 2, 3]), output)
    }

    func testDataFromStr() throws {
        let buf = try test.dataFromStr()
        let output = buf as Data
        XCTAssertEqual("abcd".data(using:.utf8)![0...3], output)
    }
    
    func testSendDataView() throws {
        let input: [UInt8] = [0, 1, 2, 3]
        let buf = Data(input)
        let output = try test.sendDataView(buf as NSData)
        XCTAssertEqual(buf, output)
    }

    func testRecvDataView() throws {
        let buf = try test.recvDataView() as Data
        XCTAssertEqual(Data([0, 1, 2, 3]), buf)
    }
}
